bigqueue.go

```diff
@@ -0,0 +1,221 @@
+package bigqueue
+
+import (
+	"fmt"
+	"path"
+)
+
+const (
+	cDataFileFmt  = "data_%d.dat"
+	cDataFileSize = 128 * 1024 * 1024
+)
+
+// BigQueue implements bigqueue interface using Go slices
+type BigQueue struct {
+	index     *QueueIndex
+	arenaList []*Arena
+	dir       string
+}
+
+// NewBigQueue constructs an instance of bigqueue
+func NewBigQueue(dir string) (*BigQueue, error) {
+	index, err := NewQueueIndex(dir)
+	if err != nil {
+		return nil, err
+	}
+
+	headAid, _ := index.GetHead()
+	tailAid, _ := index.GetTail()
+	arenaList := make([]*Arena, 0)
+
+	bq := &BigQueue{
+		index:     index,
+		arenaList: arenaList,
+		dir:       dir,
+	}
+
+	for i := headAid; i <= tailAid; i++ {
+		if err := bq.getNewArena(i); err != nil {
+			return nil, err
+		}
+	}
+
+	return bq, nil
+}
+
+// IsEmpty returns true when queue is empty
+func (bq *BigQueue) IsEmpty() bool {
+	headAid, headOffset := bq.index.GetHead()
+	tailAid, tailOffset := bq.index.GetTail()
+
+	return headAid == tailAid && headOffset == tailOffset
+}
+
+// Peek returns the head of the queue
+func (bq *BigQueue) Peek() ([]byte, error) {
+	aid, offset := bq.index.GetHead()
+
+	// read length
+	aid, offset, lenBytes, err := bq.readBytes(aid, offset, 8)
+	if err != nil {
+		return nil, err
+	}
+	length := int(BytesToUint64(lenBytes))
+
+	// read message
+	aid, offset, message, err := bq.readBytes(aid, offset, length)
+	if err != nil {
+		return message, err
+	}
+
+	return message, err
+}
+
+// Enqueue adds a new element to the tail of the queue
+func (bq *BigQueue) Enqueue(message []byte) error {
+	aid, offset := bq.index.GetTail()
+
+	// write length
+	var err error
+	length := uint64(len(message))
+	aid, offset, err = bq.writeBytes(aid, offset, Uint64ToBytes(length))
```

Comment:
We should utilize `WriteUint64` functions of mmap library. This will avoid double copy too.